x/wasm/keeper/handler_plugin.go

```diff
@@ -168,3 +177,33 @@ func (h IBCRawPacketHandler) DispatchMsg(ctx sdk.Context, _ sdk.AccAddress, cont
 	)
 	return nil, nil, h.channelKeeper.SendPacket(ctx, channelCap, packet)
 }
+
+var _ Messenger = MessageHandlerFunc(nil)
+
+// MessageHandlerFunc is a helper to construct simple function based message handler
+type MessageHandlerFunc func(ctx sdk.Context, contractAddr sdk.AccAddress, contractIBCPortID string, msg wasmvmtypes.CosmosMsg) (events []sdk.Event, data [][]byte, err error)
+
+func (m MessageHandlerFunc) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddress, contractIBCPortID string, msg wasmvmtypes.CosmosMsg) (events []sdk.Event, data [][]byte, err error) {
+	return m(ctx, contractAddr, contractIBCPortID, msg)
+}
+
+// NewBurnCoinMessageHandler handles wasmvm.BurnMsg messages
+func NewBurnCoinMessageHandler(burner types.Burner) MessageHandlerFunc {
+	return func(ctx sdk.Context, contractAddr sdk.AccAddress, _ string, msg wasmvmtypes.CosmosMsg) (events []sdk.Event, data [][]byte, err error) {
+		if msg.Bank != nil && msg.Bank.Burn != nil {
+			coins, err := convertWasmCoinsToSdkCoins(msg.Bank.Burn.Amount)
+			if err != nil {
+				return nil, nil, err
+			}
+			if err := burner.SendCoinsFromAccountToModule(ctx, contractAddr, types.ModuleName, coins); err != nil {
```

Comment:
Ah, it is a two step process? Only module accounts can burn.

Glad you took this on and not me. So many gotchas